hera_cal/lstbin.py

```diff
@@ -345,6 +345,10 @@ def lst_bin(data_list, lst_list, flags_list=None, nsamples_list=None, dlst=None,
                 # get real + imag flags
                 clip_flags = real_f + imag_f
 
+                # apply min_N condition as sigma_clip only checks axis size, not number of flags
+                sc_min_N = np.logical_not(f).sum(axis=0) < min_N
```

Comment:
Wouldn't it be more sensible to put this into the `sigma_clip` function?

Reply:
@jsdillon yes we could put it in, but then for this particular scenario where we clip both Re and Im components, `sc_min_N = np.logical_not(f).sum(axis=0) < min_N` would be repeated so leaving it outside the `sigma_clip` is a bit faster - if `sigma_clip` is used elsewhere, however, it would make sense to put it in the function.

Reply:
Is this step an appreciable fraction of the run-time of the function?

Reply:
Just looked at timings and it hardly makes a difference - I can include it in the `sigma_clip` function if you want? I also tested out the `sigma_clip` function and I think it has various problems:

- It modifies the input array `here` which means that there may be a difference if Re or Im is clipped first - this should be removed
- The `sigma_clip` function as it stands only works with 2D arrays where axis=0 is specified - anything else and it fails, so these arguments are a bit misleading. As we're only using it here for LST-binning we could make it simpler?

Reply:
yes sigma_clip purposely operates in place to save memory, and I think you've pointed out something that we actually didn't consider too closely (clipping re or im first) but you'll notice that if re or im is flagged, the other is automatically flagged (for obvious reasons) so I don't actually know if the order makes a difference. correct me if you think this is incorrect, but its a nice spot regardless.

Reply:
about the sigma_clip function args: I agree that its possible it is fairly brittle, and if you've noticed that setting axis!=0 makes it fail, then we should either fix it or get rid of the argument and hardcode axis=0. Generally I'd say we do the former, but b/c this function is very specific to the LST binner (e.g. I wouldn't use hera_cal.lstbin.sigma_clip for any generic sigma clipping routine in my analysis), I think doing the latter is a good solution: easier and less bug-prone.

Reply:
The way that `sigma_clip` is currently implemented with the function modifying the data array in place, the ordering does make a difference here. If a point is clipped for the Re component and subsequently flagged, then that point is not included in the Im sigma clipping calculations, which would change the Im MAD. We need both Re and Im clipping routines to be done on the same data array.

Reply:
Yes `sigma_clip` is brittle - I'll change it to get rid of the `axis` arg and also the `flag` arg (see below comment).

Reply:
thanks for updating the axis arg.

about the ordering, here is how this looks on master

```
# clip real
real_f = sigma_clip(d.real, sigma=sigma, min_N=min_N, axis=0)

# clip imag
imag_f = sigma_clip(d.imag, sigma=sigma, min_N=min_N, axis=0)

# clip flags
clip_flags = real_f + imag_f
```
in this case, even though d.real and d.imag are updated in place, I don't think that changes their respective flagging. notice that we don't input flags into `sigma_clip`.

lmk if i'm missing something here, thanks

Reply:
Ah, I see.. yes you're right.

I thought for some reason that `array[clip] *= np.nan` in `sigma_clip` was `nan`'ing both components of the `d` array, but in fact, it only operates on the specified component...

Just tried it for
```
a = np.ones(10, dtype=complex)
a.real[5] *= np.nan
a
```
and you do get
```
array([ 1.+0.j,  1.+0.j,  1.+0.j,  1.+0.j,  1.+0.j, nan+0.j,  1.+0.j, 1.+0.j,  1.+0.j,  1.+0.j])
```

so if Re is done first and some Re points are clipped, the data array is modified for Re only but the Im points that were originally unflagged before `sigma_clip(d.real, ...)` are still all there for `sigma_clip(d.imag, ...)`

apologies for this, understood now

Reply:
no worries! its always refreshing to get new eyes on the code--you've certainly caught a fair share of bugs here :).

so from a code perspective, could we revert changes back to master that attempted to address this?

Reply:
@nkern from a code perspective, I think the PR is still good to go, as the `array[clip] *= np.nan` in `sigma_clip` was unnecessary since `d[clip_flags] *= np.nan` is done on the outside after the clip flags are found. What do you think?

Reply:
so yes, I agree that b/c we do `array[clip_flags] *= np.nan` in the lst binner that technically the inplace operation in `sigma_clip` is irrelevant, so I'm fine w/ that.

I see you deprecated the `axis` kwarg, which I'm fine with, however, in deprecating the `flags` input I think you removed a functionality that (while we currently do not actually use) could be useful in the future. namely, in preventing the `location` variable  in `sigma_clip` from being influenced by known bad nights, for example.
now b/c the flags operation edits `array` in place, I think we might as well just keep the in-place operations in `sigma_clip` for consistency.

Reply:
ah you know what I see this is actually done in the lst binner before the `sig_clip` if statement, so nvm, I'm actually fine w/ you deprecating both the inplace operations and `flags` and `axis` kwargs in `sip_clip`, thanks!